### PR TITLE
Add setting to show/hide completion status while solving.

### LINKED ISSUE #28: Feature request: "tournament mode"
I'd like to add a "tournament mode" to XWord. When enabled, the completion status is kept hidden while solving. The puzzle is not checked (and the timer stopped) until a check/reveal is done manually.

The idea here is to practice solving in [tournament](http://www.crosswordtournament.com/) conditions, where a solution must be submitted to the judges without knowing whether every square has been filled in correctly (or at all). XWord (and other apps) currently tell you when you've completed the puzzle or else whether the grid has an error.

Without having explored the Lua scripting support much, this feels like a better fit for a script than for XWord itself since it's a bit specialized, but will have to explore whether the needed hooks are in place.

### PR DESCRIPTION
**Pull request by jpd236**
Allows solvers to hide the percent complete and correct/incorrect
notifications, as well as prevent the timer from automatically
stopping when the grid is completed correctly. This offers a more
paper-like solving experience.

Fixes #28

### CODE REVIEW

**Comment by mrichards42** [comment]
This looks good. What do you think about having the "Check Grid" button stop the timer and report the solving stats? That sounds like it might complete "tournament mode".

While testing this out I noticed some odd behavior where hitting "Cancel" in the prefs dialog resets that option. I think it's caused by the prefs dialog being static (weird move, past self), and loading the fallback config when the dialog is constructed instead of when its shown.

**Comment by jpd236** [comment]
I could do that - for me this ended up being simpler (both to implement and to describe) than what I originally proposed in the bug, and accomplished pretty much exactly what I was after. Since Check Grid doesn't have any keyboard shortcuts, I find I'd rather stop the timer first and then hit check, and I figured a setting named "tournament mode" would be mysterious, and one named "Hide completion status until you hit check" would just be a mouthful :)

But if you feel it'd be better that way, let me know and I'll give it a shot. Do you have any preferences for how we name it in that case?

I assume the odd behavior you saw was on Windows? I was testing on Mac and figured it should be the same on Windows, but maybe I missed something - will take a closer look.

**Comment by mrichards42** [comment]
TL;DR I have a few thoughts, but I don't have a great solution, so I'm happy to merge this as-is.

Compared to the other preferences "show completion status" takes a lot of explaining, though I agree that "tournament mode" would take even more explaining. That said, "tournament mode" might mean something clear to people who would use it. Both "show completion status" and "tournament mode" encompass multiple ideas in one setting (hide/show the solving percent; hide/show the correct/incorrect indicators; stop the timer when the puzzle is correct), where most of the other prefs are more granular. A full "tournament mode" probably also means that check/reveal should also be disabled, and instead you have a "done" button that stops the timer and checks the grid.

> Since Check Grid doesn't have any keyboard shortcuts [...]

Is that not mapped to `Cmd+Shift+G` for you?

> I assume the odd behavior you saw was on Windows? I was testing on Mac and figured it should be the same on Windows, but maybe I missed something - will take a closer look.

I think what's going on is that `m_oldConfig` is [initialized in the constructor](https://github.com/mrichards42/xword/blob/f42427858baff1bcd2ec495796349e3350d364ae/src/dialogs/PreferencesDialog.cpp#L49-L50), but the dialog is a singleton, and as far as I can tell `m_oldConfig` is never updated again. So when [`ResetConfig()`](https://github.com/mrichards42/xword/blob/f42427858baff1bcd2ec495796349e3350d364ae/src/dialogs/PreferencesDialog.cpp#L175) is called (from the Cancel button, which it looks like is only on Windows) it copies the stale config. I'm not 100% sure that's all right without digging into the code more, but I think that's what's going on. If so that's definitely a bug, and I'm surprised its hung around so long, but it obviously has nothing to do with the code in this PR

**Comment by jpd236** [comment]
Agreed re: the potential confusion of having one setting control multiple options. When I was implementing it I did at first envision two settings: one to show/hide the puzzle completion status indicator, and one to stop the timer upon successful solve. But I decided to group them together because stopping the timer upon successful solve is just another way to effectively see the puzzle completion status, so it felt redundant.

Regarding replacing check/reveal with done - I might find this a bit annoying, because sometimes I work on harder puzzles without worrying about time and want to be able to check/reveal if I get stuck without having to go into settings and turn off tournament mode. But maybe that's fine if "Done" causes check/reveal to be subsequently enabled?

Maybe "expert mode" would be a bit clearer, if we decide to go with that... I think it might make more intuitive sense that this turns off some of the assisting features. (And you could consider a "Beginner mode" where solution checking while typing is enabled?)

In any event, I leave it to you; I'm happy to keep this relegated to my local builds for now as it accomplishes what I was after, and/or tweak it if you've got a clear idea in mind for how you'd like it to look when merged. (Though if it gets complicated, I might not be able to get to it for a while).

You're right re: a check grid shortcut; I was thinking of the reveal options, which don't have shortcuts. And noted regarding the settings dialog - I thought your initial report was somehow specific to this new option, but it sounds like it is across the board (and definitely should be fixed either way, but probably separately).

**Comment by jpd236** [comment]
Sent out #108 to fix the problem you pointed out with cancelling on Windows. It does seem to behave exactly as you described.

I wanted to poke this again to see what your thoughts were on merging it. It seemed like you felt that "Show completion status" was a reasonable enough setting to add, even if it didn't go all the way that a "Tournament mode" might. I think it's reasonably scoped enough that it's clear what it will do, and it goes a long way towards making it feel like you're solving on paper, which is what I was after here.

As we're getting (hopefully) close to a new official release, I thought this might be good to include. I was thinking of seeing if the folks at https://www.boswords.org/ might be interested in giving it a try - they currently have a final round which uses Across Lite, which gives no way of turning off the "happy pencil".

**Comment by mrichards42** [comment]
I'm on board with this for sure, and this implementation works for me. A couple thoughts before I merge this -- not blocking, just curious to get your take on them.

1. Do you think it's better to phrase this in the positive "Show completion status" and default to true; or would it make more sense as an opt-in "Hide completion status" that defaults to false?
2. Relatedly, is this "show completion status", or "show _solving_ status"? I don't remember at this point if it also hides the solving percentage status line, or if it just prevents the green/red indicator from showing when the puzzle is done. If not, should this pref hide the solving percent as well?
3. I think this pref should disable any other prefs that it makes irrelevant (I think it's just "check solution while typing"). I wonder if it would make sense to move this checkbox above "check solution while typing", and indent the latter to make that relationship clear (like the indents in the cursor movement section). I'm not sure if that makes sense, just a thought that occurred to me.

And actually, on 3 -- unfortunately it looks like "check solution while typing" is set on the grid, not on the frame, so we probably either need to add `m_showCompletionStatus` to XGridCtrl, or automatically toggle "check solution while typing" off when the user checks "show completion status", neither of which are very pretty solutions.

**Comment by jpd236** [comment]
1. I don't feel super strongly, but I think the double negative you implicitly have here ("don't hide completion status") is slightly harder to reason about as a user. Kind of like http://truelogic.org/wordpress/2009/02/27/how-not-to-name-a-boolean-variable/ but with English instead of variables :)

2. It hides the status line entirely. But I think that still qualifies as "completion" status - the status shows the percent of the puzzle that's complete. "Show solving status" sounds fine to me too - no strong preference.

3. It's an interesting point, but ultimately I feel like we should just let users configure things how they want rather than try to guess for them. It might be unlikely, but there's nothing functionally broken about having your solution checked while typing but hiding the completion status. Maybe you just find having it visible at all times distracting. It's sort of why I leaned towards this approach rather than trying to build a more comprehensive "tournament mode" - it's simpler and more flexible to reason about showing/hiding different components of the UI than to try to come up with a tournament mode that works exactly the way everyone expects.

**Comment by mrichards42** [comment]
Thanks for the thoughts -- that's all fine by me.
